### ISSUE TITLE
DataContext.run_checkpoint() issues with RuntimeBatchRequest and batch_data with ge_cloud_id

### ISSUE THREAD

**Issue by anthonyburdi** [fivetran/great_expectations#3551]
**Describe the bug**
I'm encountering issues when trying to run a checkpoint using a dataframe passed in at runtime either via passing `batch_data` via the `batch_request` parameter or by creating a `RuntimeBatchRequest`.

**To Reproduce**
1. Create a checkpoint
2. Create a `RuntimeBatchRequest` and pass to `context.run_checkpoint()`
```
batch_request = RuntimeBatchRequest(
    datasource_name="my_datasource",
    data_connector_name="default_runtime_data_connector_name",
    data_asset_name="df_asset_name",
    runtime_parameters={"batch_data": df},
    batch_identifiers={"df_ident": datetime.date.today().strftime("%Y-%m-%d")},
)
context.run_checkpoint(
    checkpoint_name=my_checkpoint_name, 
    batch_request=batch_request
)
```
Error:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/var/folders/ds/hn_qpp1n6y3fz28clrkfmpsr0000gn/T/ipykernel_63466/4213147396.py in <module>
----> 1 context.run_checkpoint(
      2     checkpoint_name=my_checkpoint_name,
      3     batch_request=batch_request
      4 )

~/miniconda3/envs/20211014_checkpoint_notebook/lib/python3.8/site-packages/great_expectations/data_context/data_context.py in run_checkpoint(self, checkpoint_name, template_name, run_name_template, expectation_suite_name, batch_request, action_list, evaluation_parameters, runtime_configuration, validations, profilers, run_id, run_name, run_time, result_format, ge_cloud_id, expectation_suite_ge_cloud_id, **kwargs)
   3327         # TODO mark experimental
   3328         batch_request = batch_request or {}
-> 3329         batch_identifiers = batch_request.get("batch_identifiers", {})
   3330         if self.ge_cloud_mode:
   3331             if len(batch_identifiers.keys()) == 0:

AttributeError: 'RuntimeBatchRequest' object has no attribute 'get'


```

For passing batch_data directly as a dict with runtime_parameters key:
```
context.run_checkpoint(
    checkpoint_name=my_checkpoint_name, 
    batch_request={
        "runtime_parameters": {
            "batch_data": df
        }
    }
)
```
Error:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/var/folders/ds/hn_qpp1n6y3fz28clrkfmpsr0000gn/T/ipykernel_63466/202290502.py in <module>
----> 1 context.run_checkpoint(
      2     checkpoint_name=my_checkpoint_name,
      3     batch_request={
      4         "runtime_parameters": {
      5             "batch_data": df

~/miniconda3/envs/20211014_checkpoint_notebook/lib/python3.8/site-packages/great_expectations/data_context/data_context.py in run_checkpoint(self, checkpoint_name, template_name, run_name_template, expectation_suite_name, batch_request, action_list, evaluation_parameters, runtime_configuration, validations, profilers, run_id, run_name, run_time, result_format, ge_cloud_id, expectation_suite_ge_cloud_id, **kwargs)
   3379         checkpoint_run_arguments: dict = dict(**checkpoint_config, **kwargs)
   3380 
-> 3381         return checkpoint.run(**checkpoint_run_arguments)
   3382 
   3383     def test_yaml_config(

~/miniconda3/envs/20211014_checkpoint_notebook/lib/python3.8/site-packages/great_expectations/checkpoint/checkpoint.py in run(self, template_name, run_name_template, expectation_suite_name, batch_request, action_list, evaluation_parameters, runtime_configuration, validations, profilers, run_id, run_name, run_time, result_format, expectation_suite_ge_cloud_id, **kwargs)
    275             for idx, validation_dict in enumerate(validations):
    276                 try:
--> 277                     substituted_validation_dict: dict = get_substituted_validation_dict(
    278                         substituted_runtime_config=substituted_runtime_config,
    279                         validation_dict=validation_dict,

~/miniconda3/envs/20211014_checkpoint_notebook/lib/python3.8/site-packages/great_expectations/checkpoint/util.py in get_substituted_validation_dict(substituted_runtime_config, validation_dict)
    270 ) -> dict:
    271     substituted_validation_dict = {
--> 272         "batch_request": get_runtime_batch_request(
    273             substituted_runtime_config=substituted_runtime_config,
    274             validation_batch_request=validation_dict.get("batch_request"),

~/miniconda3/envs/20211014_checkpoint_notebook/lib/python3.8/site-packages/great_expectations/checkpoint/util.py in get_runtime_batch_request(substituted_runtime_config, validation_batch_request)
    263             runtime_batch_request_dict["runtime_parameters"] = {}
    264         runtime_batch_request_dict["runtime_parameters"]["batch_data"] = batch_data
--> 265     return batch_request_class(**runtime_batch_request_dict)
    266 
    267 

TypeError: __init__() got an unexpected keyword argument 'data_connector_query'
```

**Expected behavior**
I expect the checkpoint to run using the provided dataframe e.g. via RuntimeBatchRequest

**Environment (please complete the following information):**
 - MacOS
 - Great Expectations Version: 0.13.38

**Additional context**

**Comment by NathanFarmer** [comment]
Hi @anthonyburdi! Could you elaborate and/or post your checkpoint config for the checkpoint you created in step 1?

**Comment by anthonyburdi** [comment]
Great question @NathanFarmer - it is a `SimpleCheckpoint` modified from the sample in the CLI `checkpoint new` notebook:
```
my_checkpoint_name = "my_checkpoint" # This was populated from your CLI command.

yaml_config = f"""
name: {my_checkpoint_name}
config_version: 1.0
class_name: SimpleCheckpoint
run_name_template: "%Y%m%d-%H%M%S-my-run-name-template"
validations:
  - batch_request:
      datasource_name: my_datasource
#      data_connector_name: default_inferred_data_connector_name
      data_connector_name: default_runtime_data_connector_name
      data_asset_name: Titanic.csv
#      data_connector_query:
#        index: -1
    expectation_suite_name: Titanic.csv.warning
"""
print(yaml_config)
```

**Comment by liianghuang** [comment]
not sure whether this the same issue, when I use `data_context.run_checkpoint` with `batch_request=None`, I'll get Keyerror from this line
```datasource: Datasource = cast(Datasource, self.datasources[datasource_name])```

**Comment by hec10r** [comment]
Any updates on this?

**Comment by NathanFarmer** [comment]
@liianghuang @hec10r You should be able to bypass this bug by passing your `RuntimeBatchRequest` into `validations`, but in the meantime I am working on a fix for passing it into the top-level `batch_request` parameter.

**Comment by piedpipperer** [comment]
In my case, after running into some trial and error on EMR, managed to make it work the following way:

- fine tunning the spark config
- initiating the context based on files + printing the 'get_datasource' method to make sure its correctly defined. 
- Somewhat this get_datasource starts sparksession, which after that I override  with my hive and other confs. 
- At this point I read the dataframe out of the glueCatalog. and get until the validator, which is validated.
- From then, which I thought was hardest part: I am not able to run checkpoint, neither in local, will do a bit more of investigation and try to open an issue.

**Comment by NathanFarmer** [comment]
Hi @liianghuang @hec10r @piedpipperer! This issue will be resolved by one of the following PRs:
- #3680 
- #3713 
- #3718 
- #3723 

The fix will be included as part of GE release 0.13.44 later this week.
